File: android/app/src/main/java/com/phicdy/totoanticipation/model/scheduler/DeadlineAlarm.java
package com.phicdy.totoanticipation.model.scheduler;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.util.Log;

import java.util.Calendar;
import java.util.Date;

public class DeadlineAlarm {

    private final Context context;
    static final String ACTION = "deadlineAlarm";

    public DeadlineAlarm(@NonNull Context context) {
        this.context = context;
    }

    public void set5hoursBefore(@NonNull Date deadline) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(deadline);
        calendar.add(Calendar.HOUR_OF_DAY, -5);
        AlarmManager alm = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Log.d("Toto alarm", "Set alarm : " + calendar.getTime().toString());
        alm.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), generateIntent());
    }

    public void cancel() {
        AlarmManager alm = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        alm.cancel(generateIntent());
    }

    private PendingIntent generateIntent() {
        Intent i = new Intent(context, DeadlineAlarmReceiver.class);
        i.setAction(ACTION);
        return PendingIntent.getBroadcast(context, 0, i, PendingIntent.FLAG_UPDATE_CURRENT);
    }
}

File: android/scheduler/src/main/java/com/phicdy/totoanticipation/scheduler/DeadlineAlarm.kt
package com.phicdy.totoanticipation.scheduler

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.util.Log

import java.util.Calendar
import java.util.Date

class DeadlineAlarm(private val context: Context) {

    fun setAtNoonOf(deadline: Date) {
        val calendar = Calendar.getInstance().apply {
            time = deadline
            set(Calendar.HOUR_OF_DAY, 12)
            set(Calendar.MINUTE, 0)
        }
        val now = System.currentTimeMillis()
        if (now >= calendar.timeInMillis) return
        val alm = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager?
        Log.d("Toto alarm", "Set alarm : " + calendar.time.toString())
        alm?.set(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, generateIntent())
    }

    fun cancel() {
        val alm = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager?
        alm?.cancel(generateIntent())
    }

    private fun generateIntent(): PendingIntent {
        val i = Intent(context, DeadlineAlarmReceiver::class.java).apply {
            action = ACTION
        }
        return PendingIntent.getBroadcast(context, 0, i, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
    }

    companion object {
        internal const val ACTION = "deadlineAlarm"
    }
}
